src/c++/uds/kernelLinux/tests/getTestIndexNames.c

```diff
@@ -31,83 +31,125 @@
 
 enum { BLK_FMODE = FMODE_READ | FMODE_WRITE };
 
+struct block_device_context {
+	struct block_device *block_device;
+#ifdef VDO_USE_ALTERNATE
+#ifndef VDO_USE_ALTERNATE_2
+	struct bdev_handle *device_handle;
+#endif /* VDO_USE_ALTERNATE_2 */
+#else
+	struct file *file_handle;
+#endif /* VDO_USE_ALTERNATE */
+};
+
+static struct block_device_context contexts[2];
+
 /**********************************************************************/
-static struct block_device *get_device_from_name(const char *name)
+static void set_device_context(int context_number, const char *name)
 {
-	struct block_device *bdev;
-
+	struct block_device_context *context = &contexts[context_number];
 #ifdef VDO_USE_ALTERNATE
 #ifdef VDO_USE_ALTERNATE_2
 #ifdef VDO_USE_ALTERNATE_3
 
-	bdev = blkdev_get_by_path(name, BLK_FMODE, NULL);
+	context->block_device = blkdev_get_by_path(name, BLK_FMODE, NULL);
 #else
 	const struct blk_holder_ops hops = { NULL };
 
-	bdev = blkdev_get_by_path(name, BLK_FMODE, NULL, &hops);
+	context->block_device = blkdev_get_by_path(name, BLK_FMODE, NULL, &hops);
 #endif /* VDO_USE_ALTERNATE_3 */
 #else
 	const struct blk_holder_ops hops = { NULL };
```

Comment:
At least some flavors of the bdev open functions save the "hops" pointer away. I think the structures need to be "static const" if they're defined within the function body, or moved to a top-level definition, because they'll be on the stack otherwise.

Reply:
I don't think we care whether the hops are on the stack, because they are under the caller's control (i.e., our control). The problem with the device_handle is that the kernel controls the memory, so it tries to free it. But with the hops, that doesn't happen.

I could move the hops into the block_device_context, but I really think it can't matter.

Reply:
There are calls into the holder ops from bdev.c, I'm not convinced we fully control it all.
Since the values are constants, I don't think they need to move, they just need static storage allocation. Though, from other examples in the kernel, I think passing a null pointer may be allowed too.

Reply:
Okay, fine, I've added the static.